### PR TITLE
Fixing edge connectivity for OSM ways with differing layer tag values

### PR DESCRIPTION
**Pull request by MikeGost**
### Description:

In the existing implementation, we would only way-section two ways if they had the same layer tag value. However, this ignored a very common OSM use-case where two ways could be connected, but have two distinct layer tag values. As an example, imagine a trunk-link of layer 0 linking a trunk road of layer 1 to another road of layer 0 - a very common case for most on/off ramps. This PR updates the way-sectioning logic to allow sectioning if two ways have different layer tag values, as long as the intersection is at the end point of one of the ways. This is a clear distinction, to avoid sectioning anything that is intersecting at a non-endpoint, since this is often times a mapping error. As an example, imagine an overpass road of layer 1 sharing a shape point with an underlying residential road of layer 0. If we way-section at a non-end point, we would end up connecting these two, when in reality there is no way to drive from the overpass to the residential road beneath it. It's worth calling out - that the existing logic may still have some corner cases that exist. However, these changes align the Atlas more closely with the OSM data model. 

### Potential Impact:

This will have an impact on any consumer of the atlas with the changes. Specifically, the proper connectivity is now ensured with ways that are connected at endpoints and having different layer tag values.

### Unit Test Approach:

Added 2 new unit tests and renamed an existing one to capture behavior more accurately. The test suite is now comprised of three cases where two ways with different layer tag values are intersecting. The intersections occur at the beginning, middle and end of the way. When the intersection happens at the start or end of the way, the proper connectivity is ensured. Whereas, if the intersection is during a non-end point between the two ways of different layer tag values, then we ensure they are not connected. The two unit tests both fail without the changes and pass afterwards. Below are the screenshots of the behavior the unit tests are testing:

Intersection at start of the way:
intersectionatstart

Intersection at end (and also at the end) of the way:
intersectionatend

### Test Results:

Ran the changes on a shard that contains such intersections and visually verified correct stitching and connectivity.
------

In doubt: [Contributing Guidelines](CONTRIBUTING.md)

### CODE REVIEW

File: src/main/java/org/openstreetmap/atlas/geography/atlas/raw/sectioning/WaySectionProcessor.java [hunk 1]
```diff
@@ -779,6 +788,42 @@ private boolean isAtlasPoint(final WaySectionChangeSet changeSet, final Point po
...
+     */
+    private boolean locationIsAnEndPointOfAnIntersectingEdgeOfDifferentLayer(
+            final Location location, final Line line)
+    {
+        final long targetLayerValue = LayerTag.getTaggedOrImpliedValue(line, 0L);
+
+        return Iterables
+                // Find all intersecting edges
+                .stream(this.rawAtlas.linesContaining(location,
+                        target -> target.getIdentifier() != line.getIdentifier()
```
[the remark this answers is not in the document]
Comment (matthieun) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/raw/sectioning/WaySectionProcessor.java]: This way-sectioning phase works on a "raw" atlas that only contains lines, points and relations. This step is what determines what becomes an edge and a line, and that creates the reverse edges. At that point, there are no reverse edges yet.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/raw/sectioning/WaySectionProcessor.java [hunk 2]
```diff
@@ -779,6 +788,42 @@ private boolean isAtlasPoint(final WaySectionChangeSet changeSet, final Point po
...
+                                && isAtlasEdge(target) && target.asPolyLine().contains(location)))
```
[the remark this answers is not in the document]
Comment (matthieun) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/raw/sectioning/WaySectionProcessor.java]: It only looks at the shape points. But @MikeGost can correct me here, I think `location` is only a start/end point, so it should match with a shape point, right?

File: src/main/java/org/openstreetmap/atlas/geography/atlas/raw/sectioning/WaySectionProcessor.java [hunk 3]
```diff
@@ -625,11 +627,18 @@ private void identifyEdgesNodesAndAreasFromLines(final WaySectionChangeSet chang
...
                     }
 
-                    // 3. Check if there is an edge intersection of the same layer at this location
+                    // 3. Check if there is an intersecting edge, of the same layer at this location
                     if (locationIsPartOfAnIntersectingEdgeOfTheSameLayer(shapePoint, line))
                     {
                         addPointToNodeList(shapePoint, nodesForEdge);
                     }
+
+                    // 4. Check if there is an intersecting edge, of a different layer at this
```
[the remark this answers is not in the document]
Comment (MikeGost) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/raw/sectioning/WaySectionProcessor.java]: Good point, I'll update the comment to reflect the 3b

File: src/main/java/org/openstreetmap/atlas/geography/atlas/raw/sectioning/WaySectionProcessor.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (MikeGost) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/raw/sectioning/WaySectionProcessor.java]: Correction, the location is just a shape point, with a possible intersection that we are checking for.

**Comment by MikeGost** [comment]
Good point @matthieun - that was my plan for today :)

File: src/main/java/org/openstreetmap/atlas/geography/atlas/raw/sectioning/WaySectionProcessor.java [hunk 4]
```diff
@@ -779,6 +788,42 @@ private boolean isAtlasPoint(final WaySectionChangeSet changeSet, final Point po
...
+                // and that our point is its start or end node
+                .anyMatch(candidate ->
+                {
+                    final long layerValue = LayerTag.getTaggedOrImpliedValue(candidate, 0L);
+                    final boolean edgesOnDifferentLayers = targetLayerValue != layerValue;
+
+                    final PolyLine candidatePolyline = candidate.asPolyLine();
+                    final boolean intersectionIsAtEndPoint = candidatePolyline.first()
+                            .equals(location) || candidatePolyline.last().equals(location);
+                    return edgesOnDifferentLayers && intersectionIsAtEndPoint;
```
[the remark this answers is not in the document]
Comment (lucaspcram) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/raw/sectioning/WaySectionProcessor.java]: This is true, but worst case it's 2 extra byte code instructions (an `iload_` and an `istore_` for the `intersectionIsAtEndPoint` variable).

I think having the name of what that predicate represents will be useful if we need to dig back into the code in the future.

[the remark this answers is not in the document]
Comment (MikeGost) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/raw/sectioning/WaySectionProcessor.java]: I'll change it back!

[the remark this answers is not in the document]
Comment (lucaspcram) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/raw/sectioning/WaySectionProcessor.java]: Code tennis! :)

File: src/main/java/org/openstreetmap/atlas/geography/atlas/raw/sectioning/WaySectionProcessor.java [hunk 5]
```diff
@@ -779,6 +788,42 @@ private boolean isAtlasPoint(final WaySectionChangeSet changeSet, final Point po
...
+                .stream(this.rawAtlas.linesContaining(location,
+                        target -> target.getIdentifier() != line.getIdentifier()
+                                && isAtlasEdge(target) && target.asPolyLine().contains(location)))
+                // Check whether that edge has a different layer value as the line we're looking at
+                // and that our point is its start or end node
+                .anyMatch(candidate ->
+                {
+                    final long layerValue = LayerTag.getTaggedOrImpliedValue(candidate, 0L);
+                    final boolean edgesOnDifferentLayers = targetLayerValue != layerValue;
+                    final PolyLine candidatePolyline = candidate.asPolyLine();
```
[the remark this answers is not in the document]
Comment (MikeGost) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/raw/sectioning/WaySectionProcessor.java]: Gave this some thought during original implementation. I don't think the performance improvement would be that great, considering that the intersection logic is fairly trivial (two equality checks for Location class). My thought was to weight readability over minor performance. Will consider changing if I see this misbehaving on the cluster :)

File: src/main/java/org/openstreetmap/atlas/geography/atlas/raw/sectioning/WaySectionProcessor.java [hunk 6]
```diff
@@ -779,6 +788,42 @@ private boolean isAtlasPoint(final WaySectionChangeSet changeSet, final Point po
...
+     *            The {@link Line} to use
+     * @return {@code true} if there is at least one {@link Edge} with a different layer tag value
+     *         intersecting the given line and having an end point at the given location
+     */
+    private boolean locationIsAnEndPointOfAnIntersectingEdgeOfDifferentLayer(
+            final Location location, final Line line)
+    {
+        final long targetLayerValue = LayerTag.getTaggedOrImpliedValue(line, 0L);
+
+        return Iterables
```
[the remark this answers is not in the document]
Comment (MikeGost) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/raw/sectioning/WaySectionProcessor.java]: Same comment here. Going to keep as is for readability here.

**Pull request merged**